### PR TITLE
refactor __init__

### PR DESCRIPTION
**Pull request by jwindhager**
Hi @JoOkuma,

I found `BaseGraph.__init__` a bit hard to parse and attempted to refactor it:
- `__init__` is structured into validation and initialization, where validation also infers dimensions
- `__init__` now allows specifying all arguments (including `coords` + `ndim`), as long as everything matches
- `_init_buffers` has been split up into `_init_node_buffers` and `_init_edge_buffers`

Just let me know what you think - also happy to revert back to your original version if you don't like it!

### CODE REVIEW

File: src/napari_graph/_base_graph.py [hunk 1]
```diff
@@ -190,54 +190,73 @@ def __init__(
...
+            if n_edges < len(edges):
                 raise ValueError(
-                    f"`n_nodes` ({n_nodes}) must be greater or equal than `coords` length ({len(coords)})."
+                    f"`n_edges` ({n_edges}) must be greater or equal than `edges` length ({len(edges)}."
                 )
+        else:
+            n_edges = len(edges)
 
-        elif n_nodes is None:
+        # initialize nodes
```
Comment (JoOkuma) [inline src/napari_graph/_base_graph.py]: I would move  "initialize nodes" with the section "validate nodes". This could make some things clearer, for example, when `self._coords` is set which uses the fact that `ndim` was set during validation.

File: src/napari_graph/_base_graph.py [hunk 2]
```diff
@@ -190,54 +190,73 @@ def __init__(
...
+        if n_nodes is None:
             n_nodes = 0
```
Comment (JoOkuma) [inline src/napari_graph/_base_graph.py]: I think it might be interesting to set `n_nodes` and `n_edges` to `_ALLOC_MIN` when no value was found. So the buffer isn't empty when the user starts adding nodes/edges.

**Comment by JoOkuma** [comment]
I left some comments above, but it already looks much clearer.

File: src/napari_graph/_base_graph.py (hunk 2 above)
Comment (jwindhager) [reply src/napari_graph/_base_graph.py]: Good point! Although I think this only makes sense for nodes, as `n_edges` is inferred from the mandatory `edges` argument. I've added another commit for this.

**Comment by jwindhager** [comment]
Sorry, accidentally hit the wrong button - any more input on this PR?

**Comment by JoOkuma** [comment]
Looks good to me!
